bluephos/tasks/dft.py

```diff
@@ -0,0 +1,99 @@
+from dplutils.pipeline import PipelineTask
+
+# Setup logging and get a logger instance
+logger = log_config.setup_logging(__name__)
+
+# DEBUG = True retains DFT details for review; False only keeps final results
+DEBUG = True
+
+# Constants
+MAX_DEFAULT_CPUS = 48  # Maximum default CPUs to use if not specified by environment
+
+
+def get_dft_calculator(package, n_cpus):
+    """
+    Get the appropriate DFT calculator based on the specified package.
+
+    Args:
+        package (str): The name of the DFT package to use ('orca' or 'ase').
+        n_cpus (int): Number of CPUs to use for the calculations.
+
+    Returns:
+        An instance of the specified DFT calculator.
+
+    Raises:
+        ValueError: If an unsupported DFT package is specified.
+    """
+
+    if package == "orca":
+        orca_path = os.path.join(os.getenv("EBROOTORCA"), "orca")
+        return OrcaCalculator(n_cpus, orca_path)
+    elif package == "ase":
+        return ASECalculator(n_cpus)
+    else:
+        raise ValueError("Unsupported DFT package")
+
+
+# Process each row of the DataFrame to perform DFT calculations
+def process_dataframe(row, t_ste, dft_calculator):
+    mol_id = row["ligand_identifier"]
+    ste = row["ste"]
+    energy_diff = row["energy diff"]
+
+    if ste is None or abs(ste) >= t_ste or energy_diff is not None:
+        logger.info(f"Skipping DFT on molecule {mol_id} based on z or t_ste conditions.")
+        return row
+
+    if row["xyz"] not in ["failed", None]:
+        base_name = row["ligand_identifier"]
+    else:
+        return row
+
+    if DEBUG:
+        temp_dir = tempfile.mkdtemp()
+    else:
+        temp_dir_obj = tempfile.TemporaryDirectory()
+        temp_dir = temp_dir_obj.name
+
+    try:
+        xyz_value = remove_second_row(row["xyz"])
+        logger.info(f"Starting DFT calculation for {base_name}...")
+        energy_diff = dft_calculator.extract_results(temp_dir, base_name, xyz_value)
+        row["energy diff"] = energy_diff
+        return row
+    finally:
+        if DEBUG:
+            logger.info(f"Temporary files for {base_name} kept at {temp_dir} for debugging.")
+        else:
+            if isinstance(temp_dir, tempfile.TemporaryDirectory):
+                temp_dir.cleanup()
+
+
+# Run DFT calculations on the DataFrame
+def dft_run(df: pd.DataFrame, t_ste: float, package: str) -> pd.DataFrame:
+    n_cpu_custom = min(multiprocessing.cpu_count(), MAX_DEFAULT_CPUS)
+    n_cpus = int(os.getenv("OMP_NUM_THREADS", str(n_cpu_custom)))
+
+    dft_calculator = get_dft_calculator(package, n_cpus)
+
+    df = df.apply(process_dataframe, axis=1, t_ste=t_ste, dft_calculator=dft_calculator)
+    return df
+
+
+# Define the PipelineTask for the DFT run
+DFTTask = PipelineTask(
+    "dft_run",
+    dft_run,
+    context_kwargs={
+        "t_ste": "t_ste",
+        "package": "package",  #  Either "ase" or "orca"
+    },
+    batch_size=1,
+    num_cpus=64,
```

Comment:
That's an important setting for the DFT calculation. Here are my thoughts:
1) In Niagara, if the maximum number of CPUs is 40, we can set it to 40 to fully utilize the server.
2) I experimented and found that the DFT task performs better with fewer CPUs. For example:
64 CPUs: 23 minutes
48 CPUs: 16 minutes
32 CPUs: 19 minutes
24 CPUs: 22 minutes
16 CPUs: 33 minutes
3)Assuming the optimal number of CPUs for one ligand is 40 (15 minutes), and for 20 CPUs it's 25 minutes, we should define num_cpus = 40 and batch_size = 2.
4)In addition, we could make num_cpus, num_gpu, and batch_size configurable via the command line or the same file as the thresholds ? This would centralize adjustments and simplify changes.

Reply:
those settings are already easily configurable via command line, we just need to set reasonable defaults. Actually from the looks of the above, it might be better to sacrifice some individual runtime to pack more instances on a single node. But we could probably do well by collected a larger amount of data to better characterize this.

Given the runtimes, batch size of 1 by default probably makes best sense here

Reply:
Oh, yes! If the batch size is 2, that means only 1 batch (2/2=1) is running on the cluster, with each ligand in this batch utilizing all the CPUs.

Reply:
or to be a bit more precise: only one DFT running in a task taking 64 CPUs at a time. A cluster with more cores could run more in parallel (e.g. 128 cores could run 2 simultaneously). 

So far it is reading to me as a lower number of cores and more simultaneous tasks is a good tradeoff since the scaling is far from linear. But you are correct that this is config, so can be tuned to needs at runtime!